portal/src/views/shared/Comments.vue

```diff
@@ -0,0 +1,371 @@
+                                <UserPhoto :user="user"> </UserPhoto>
+                                <input type="text" placeholder="Reply to comment" v-model="newReply.body">
+                                <button type="submit" class="new-comment-submit" v-if="newReply.body"> Post </button>
+                            </form>
+                        </transition>
+
+                        <div class="actions">
+                            <button @click="addReply(post)">
+                                <img src="@/assets/icon-reply.svg">
+                                Reply
+                            </button>
+                            <button v-if="viewType === 'data'" @click="viewDataClick(post)">
+                                <img src="@/assets/icon-view-data.svg">
+                                View Data
+                            </button>
+                        </div>
+                    </div>
+                </div>
+            </transition-group>
+        </div>
+    </section>
+</template>
+
+<script lang="ts">
+
+import Vue from "vue";
+import CommonComponents from "@/views/shared";
+import moment from "moment";
+import {Bookmark} from "@/views/viz/viz";
+
+export interface Comment {
+    id: number,
+    author: object,
+    body: string;
+    replies: [];
+    createdAt: number;
+    updatedAt: number;
+}
+
+export default Vue.extend({
+    name: "Comments",
+    components: {
+        ...CommonComponents,
+    },
+    props: {
+        user: {
+            required: true,
+        },
+        parentData: {
+            required: true,
+            type: [Number, Bookmark],
+        },
+    },
+    methods: {
+        getNewCommentPlaceholder(): string {
+            if (this.viewType === "project") {
+                return "Comment on Project";
+            } else {
+                return "Write a comment about this Data View";
+            }
+        },
+        save(comment): void {
+            this.$services.api
+                .postComment(comment)
+                .then((response: {post: Comment}) => {
+                    this.newComment.body = null;
+                    // add the comment to the replies array
+                    if (comment.threadId) {
+                        this.posts.filter((post) => post.id === comment.threadId)[0].replies.push(response.post);
+                        this.newReply.body = null;
+                    } else {// add it to the posts array
+                        this.posts.unshift(response.post);
+                        this.newComment.body = null;
+                    }
+                }).catch(e => {
+                    this.errorPostComment = true;
+                });
+        },
+        formatTimestamp(timestamp): string {
+            return moment(timestamp).fromNow();
+        },
+        addReply(post): void {
```

Comment:
I'd also like to see types on the parameters. This helps ensure functions conform to the stated types being passed in.

Reply:
done